h5py/tests/test_objects.py

```diff
@@ -34,3 +42,91 @@ def test_hash(self):
         oid = o.ObjectID(42)
         with self.assertRaises(TypeError):
             hash(oid)
+
+    def test_phil_fork_with_threads(self):
+        # Test that handling of the phil Lock after fork is correct
+        # even when multiple threads are present.
+
+        # On Windows forking (and the register_at_fork handler)
+        # are not available, skip this test.
+        if not hasattr(os, "fork"):
+            raise SkipTest("os.fork not available")
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            fns = []
+            for i in range(10):
+                fn = os.path.join(tmpdir, f'test{i}.h5')
+                with h5py.File(fn, 'w') as f:
+                    f.create_dataset('values', data=np.random.rand(1000, 1000))
+                fns.append(fn)
+
+            def f(fn):
+                with h5py.File(fn, 'r') as f:
+                    for _ in range(100):
+                        _ = f['values'][:]
+
+            # create 10 threads, each reading from an HDF5 file
+            threads = []
+            for fn in fns:
+                thread = threading.Thread(target=f, args=(fn,))
+                thread.start()
+                threads.append(thread)
+
+            # While the threads are running (and potentially holding the phil Lock)
+            # create 10 processes, each also reading from an HDF5 file
+            worker2pid = {}
+            for worker_id, fn in enumerate(fns):
+                pid = os.fork()
+                if pid == 0:
+                    # child process
+                    f(fn)
+                    os._exit(0)
+                else:
+                    # parent process
+                    worker2pid[worker_id] = pid
+
+            # Wait for all child processes to finish
+            for worker_id, pid in worker2pid.items():
+                os.waitpid(pid, 0)
+
+            # Wait for all threads to finish
+            for thread in threads:
+                thread.join()
+
+    def test_phil_fork_with_threads_2(self):
+        # Test that handling of the phil Lock after fork is correct.
+        # We simulate a deadlock in the forked process by explicitly
+        # waiting for the phil Lock to be acquired in a different thread
+        # before forking.
+
+        # On Windows forking (and the register_at_fork handler)
+        # are not available, skip this test.
+        if not hasattr(os, "fork"):
+            raise SkipTest("os.fork not available")
+
+        thread_acquired_phil_event = threading.Event()
+
+        def f():
+            o.phil.acquire()
+            try:
+                thread_acquired_phil_event.set()
+                time.sleep(1)
+            finally:
+                o.phil.release()
+
+        thread = threading.Thread(target=f)
+        thread.start()
+        try:
+            # on the main thread wait for the thread to have acquired the phil lock
```

Comment:
this comment + your naming convention is confusing to me: which one is the "main" thread ? variable names would lead me to conclude that's `thread`, but the logic as a whole appears to imply the opposite.